File: elisp/load.cc
// Copyright 2020, 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "elisp/load.h"

#include <iostream>
#include <string>
#include <vector>

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
#pragma GCC diagnostic ignored "-Wconversion"
#pragma GCC diagnostic ignored "-Wsign-conversion"
#pragma GCC diagnostic ignored "-Woverflow"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/utility/utility.h"
#include "tools/cpp/runfiles/runfiles.h"
#pragma GCC diagnostic pop

#include "elisp/process.h"
#include "elisp/status.h"

namespace phst_rules_elisp {

absl::Status AddLoadPath(const Runfiles& runfiles,
                         std::vector<std::string>& args,
                         const std::vector<std::string>& load_path) {
  constexpr const char* const runfiles_elc =
      "phst_rules_elisp/elisp/runfiles/runfiles.elc";
  bool runfile_handler_installed = false;
  for (const auto& dir : load_path) {
    const auto status_or_dir = Runfile(runfiles, dir);
    if (status_or_dir.ok()) {
      args.push_back(absl::StrCat("--directory=", status_or_dir.value()));
    } else if (absl::IsNotFound(status_or_dir.status())) {
      if (!absl::exchange(runfile_handler_installed, true)) {
        ASSIGN_OR_RETURN(const auto file, Runfile(runfiles, runfiles_elc));
        args.push_back(absl::StrCat("--load=", file));
        args.push_back("--funcall=elisp/runfiles/install-handler");
      }
      args.push_back(absl::StrCat("--directory=/bazel-runfile:", dir));
    } else {
      return status_or_dir.status();
    }
  }
  return absl::OkStatus();
}

}  // namespace phst_rules_elisp

File: elisp/load.py
# Copyright 2021, 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Functions for load path management.

This is an internal helper library for the Emacs Lisp Bazel rules.  Don’t rely
on it in any way outside the rule implementation."""

from collections.abc import Iterable
import pathlib

from elisp import runfiles

def add_path(run_files: runfiles.Runfiles, args: list[str],
             load_path: Iterable[pathlib.PurePosixPath]) -> None:
    """Add load path elements to the given args list."""
    runfiles_elc = pathlib.PurePosixPath(
        'phst_rules_elisp/elisp/runfiles/runfiles.elc')
    runfile_handler_installed = False
    for directory in load_path:
        try:
            args.append('--directory=' + str(run_files.resolve(directory)))
        except FileNotFoundError:
            if not runfile_handler_installed:
                file = run_files.resolve(runfiles_elc)
                args += [
                    '--load=' + str(file),
                    '--funcall=elisp/runfiles/install-handler',
                ]
                runfile_handler_installed = True
            args.append('--directory=/bazel-runfile:' + str(directory))
